sklearn/_loss/loss.py

```diff
@@ -1128,6 +1276,57 @@ def gradient_proba(
         )
         return gradient_out, proba_out
 
+    def loss_array_api(self, y_true, raw_prediction, sample_weight=None, xp=None):
+        xp, _, device_ = get_namespace_and_device(
+            y_true, raw_prediction, sample_weight, xp=xp
+        )
+        log_sum_exp = _logsumexp(raw_prediction, axis=1, xp=xp)
+        y_true = xp.asarray(y_true, dtype=xp.int64, device=device_)
+        class_margins = (
+            xp.arange(y_true.shape[0], device=device_) * raw_prediction.shape[1]
+        )
+        label_predictions = xp.take(_ravel(raw_prediction), y_true + class_margins)
+        loss = log_sum_exp - label_predictions
+        if sample_weight is not None:
+            loss *= sample_weight
+        return loss
+
+    def gradient_array_api(self, y_true, raw_prediction, sample_weight=None, xp=None):
+        xp, _, device_ = get_namespace_and_device(
+            y_true, raw_prediction, sample_weight, xp=xp
+        )
+        y_true = xp.asarray(y_true, dtype=xp.int64, device=device_)
+        y_true_one_hot = y_true[:, None] == xp.arange(
+            raw_prediction.shape[1], device=device_
+        )
```

Comment:
We should pre-allocate and reuse this kind of datastructurs in the constructor, or in a `prepare` method of the loss object and then reuse them instead of reallocating them at each iteration (assuming that the number of data points and classes do not change during a fit call).